include/exadg/solvers_and_preconditioners/preconditioners/preconditioner_amg.h

```diff
@@ -154,17 +274,201 @@ class PreconditionerML : public PreconditionerBase<double>
+    }
 
     this->update_needed = false;
   }
 
 private:
+  Teuchos::ParameterList
+  get_parameter_list()
+  {
+    Teuchos::ParameterList parameter_list;
+
+    // Slightly modified from deal::PreconditionAMG::AdditionalData::set_parameters().
+    if(ml_data.elliptic == true)
+    {
+      ML_Epetra::SetDefaults("SA", parameter_list);
+      if(ml_data.higher_order_elements)
+      {
+        parameter_list.set("aggregation: type", "Uncoupled");
+      }
+    }
+    else
+    {
+      ML_Epetra::SetDefaults("NSSA", parameter_list);
+      parameter_list.set("aggregation: type", "Uncoupled");
+      parameter_list.set("aggregation: block scaling", true);
+    }
+
+    parameter_list.set("smoother: type", ml_data.smoother_type);
+    parameter_list.set("coarse: type", ml_data.coarse_type);
+
+// Force re-initialization of the random seed to make ML deterministic
+// (only supported in trilinos >12.2):
+#  if DEAL_II_TRILINOS_VERSION_GTE(12, 4, 0)
+    parameter_list.set("initialize random seed", true);
+#  endif
+
+    parameter_list.set("smoother: sweeps", static_cast<int>(ml_data.smoother_sweeps));
+    parameter_list.set("cycle applications", static_cast<int>(ml_data.n_cycles));
+    if(ml_data.w_cycle == true)
+    {
+      parameter_list.set("prec type", "MGW");
+    }
+    else
+    {
+      parameter_list.set("prec type", "MGV");
+    }
+
+    parameter_list.set("smoother: Chebyshev alpha", 10.);
+    parameter_list.set("smoother: ifpack overlap", static_cast<int>(ml_data.smoother_overlap));
+    parameter_list.set("aggregation: threshold", ml_data.aggregation_threshold);
+
+    // Minimum size of the coarse problem, i.e. no coarser problems
+    // smaller than `coarse: max size` are constructed.
+    parameter_list.set("coarse: max size", 2000);
+
+    // This extends the settings in deal::PreconditionAMG::AdditionalData::set_parameters().
+    parameter_list.set("repartition: enable", 1);
+    parameter_list.set("repartition: max min ratio", 1.3);
+    parameter_list.set("repartition: min per proc", 300);
+    parameter_list.set("repartition: partitioner", "Zoltan");
+    parameter_list.set("repartition: Zoltan dimensions", static_cast<int>(dim));
+
+    if(ml_data.output_details)
+    {
+      parameter_list.set("ML output", 10);
+    }
+    else
+    {
+      parameter_list.set("ML output", 0);
+    }
+
+    return parameter_list;
+  }
+
+  std::vector<VectorType>
+  get_constant_modes()
+  {
+    unsigned int n_constant_modes = dealii::numbers::invalid_unsigned_int;
+    if(operator_type == AMGOperatorType::ScalarLaplace)
+    {
+      n_constant_modes = 1;
+    }
+    else if(operator_type == AMGOperatorType::VectorLaplace)
+    {
+      n_constant_modes = dim;
+    }
+    else if(operator_type == AMGOperatorType::Elasticity)
+    {
+      n_constant_modes = dim == 1 ? 1 : (dim == 2 ? 3 : 6);
+    }
+    else
+    {
+      AssertThrow(false, dealii::ExcMessage("AMGOperatorType not defined"));
+    }
```

Comment:
Related to comment https://github.com/exadg/exadg/pull/693/files#r1705187688, this seems to be duplicated code. To avoid this, I would let the class `ConstantModes<dim>` take over the responsibillity for `n_constant_modes`.  `ConstantModes<dim>` receives the operator type in the constructur and has a member function `n_modes()` to perform a loop. When looping over all the modes to compute the vectors, one calls another member function `mode_generator.set_mode(i)`, similar to `dealii::Function::set_time()`.

Reply:
its not needed anymore, thanks for your effort! :)